File: 725-split-linked-list-in-parts/split-linked-list-in-parts.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode[] splitListToParts(ListNode head, int k) {
        ListNode dummy = head;
        int count = 0;

        while (dummy != null) {
            count++;
            dummy = dummy.next;
        }
        
        int parts = count / k;
        int rest = count % k;
        ListNode[] ans = new ListNode[k];

        int i = 0;
        while (k-- > 0) {
            int t = parts;
            if (rest > 0) {
                t++;
                rest--;
            }
            ListNode newHead = head;
            ListNode current = newHead;
            while (newHead != null && t-- > 1) {
                newHead = newHead.next;
            }

            if (newHead != null) {
                head = newHead.next;
                newHead.next = null;
            }
            ans[i] = current;
            i++;
        }
        return ans;
    }
}

File: 725-split-linked-list-in-parts/split-linked-list-in-parts.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
struct ListNode** splitListToParts(struct ListNode* head, int k, int* returnSize) {
    struct ListNode* dummy = head;
    int count = 0;

    while (dummy != NULL) {
        count++;
        dummy = dummy->next;
    }

    int parts = count / k;
    int rest = count % k;
    struct ListNode** ans = (struct ListNode**)malloc(k * sizeof(struct ListNode*));
    *returnSize = k;

    int i = 0;
    while (k > 0) {
        int t = parts;
        if (rest > 0) {
            t++;
            rest--;
        }
        struct ListNode* newHead = head;
        struct ListNode* current = newHead;
        while (newHead != NULL && t > 1) {
            newHead = newHead->next;
            t--;
        }

        if (newHead != NULL) {
            head = newHead->next;
            newHead->next = NULL;
        }
        ans[i] = current;
        i++;
        k--;
    }
    return ans;
}